File: backend/app/models/check_record.py
from __future__ import annotations

import enum
from datetime import datetime, timezone

from sqlalchemy import Boolean, DateTime, Enum, Float, ForeignKey, Integer
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base


class CheckType(str, enum.Enum):
    clock_in = "clock_in"
    clock_out = "clock_out"


class CheckRecord(Base):
    __tablename__ = "check_records"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    employee_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("employees.id", ondelete="CASCADE"), nullable=False
    )
    check_time: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    check_type: Mapped[CheckType] = mapped_column(Enum(CheckType), nullable=False)
    face_passed: Mapped[bool | None] = mapped_column(Boolean, nullable=True)
    location_lat: Mapped[float | None] = mapped_column(Float, nullable=True)
    location_lng: Mapped[float | None] = mapped_column(Float, nullable=True)
    location_passed: Mapped[bool | None] = mapped_column(Boolean, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime, default=lambda: datetime.now(timezone.utc)
    )

File: frontend/app/lib/shared/models/check_record.dart
class CheckRecord {
  final int id;
  final DateTime checkTime;
  final String checkType;
  final bool? facePassed;
  final double? locationLat;
  final double? locationLng;
  final bool? locationPassed;

  const CheckRecord({
    required this.id,
    required this.checkTime,
    required this.checkType,
    this.facePassed,
    this.locationLat,
    this.locationLng,
    this.locationPassed,
  });

  factory CheckRecord.fromJson(Map<String, dynamic> json) {
    // 后端返回 UTC 时间，确保正确转换为本地时间
    var timeStr = json['check_time'] as String;
    // 如果没有时区标识，强制当作 UTC
    if (!timeStr.endsWith('Z') && !timeStr.contains('+')) {
      timeStr = '${timeStr}Z';
    }
    return CheckRecord(
      id: json['id'] as int,
      checkTime: DateTime.parse(timeStr).toLocal(),
      checkType: json['check_type'] as String,
      facePassed: json['face_passed'] as bool?,
      locationLat: (json['location_lat'] as num?)?.toDouble(),
      locationLng: (json['location_lng'] as num?)?.toDouble(),
      locationPassed: json['location_passed'] as bool?,
    );
  }

  bool get isClockIn => checkType == 'clock_in';
  bool get isClockOut => checkType == 'clock_out';
}
